### PR TITLE
[feat] Add TOTP support

### PR DESCRIPTION
**Pull request by dan-stytch**
## Summary
First and foremost, this PR add support for TOTP. It also adds a bit of a testing overhaul with respect to how the repo manages setting up mock requests (explained below), and also adds some convenience items, as listed below.

## Changes
* TOTP Support
* Add NetworkingClientInterceptor to BaseTestCase and set up `.repsonses { response1 }` resultBuilder for simpler mock response set up
* Adds TOTP to StytchDemo app
* Refactors StytchDemo app to have closures pass back single AuthenticateResponseType vs Session and User objects
* Converts some string IDs to be strongly typed
* Adds several convenience items (e.g. JSON init to empty object)
* Adds a `Union` type with dynamicMemberLookup to add additional data to existing types

Completes SDK-767, SDK-768, SDK-770

### NetworkingClientInterceptor details
The test module now provides a simple type which keeps track of requests (so you can assert against the array of requests( and allows simple setting of the mock responses via a `.responses {}` result builder function. This function will automatically serialize success and failures based on whether they are Codable or Error types (though this can be explicitly handled in the case of ambiguity.) Additionally, serialized success responses are automatically wrapped in a json container under the "data" key, for simpler ergonomics.

The syntax for using this looks something like this:
``` swift
networkInterceptor.responses {
    aSuccessfulResponse
    someError
    Failure {
        anotherError
        oneMoreError
    }
    Success {
        aSuccess
        oneMoreSuccess
    }
}
```
The above setup would cause the networking interceptor to have the following responses queued up for use:
`[.success(aSuccessfulResponse), .failure(someError), .failure(anotherError), .failure(oneMoreError), .success(aSuccess), .success(oneMoreSuccess)]`
// NOTE: the successes would, in actuality, be serialized Data objects wrapped in a json container under the "data" key. The above was typed for simplicity.

### CODE REVIEW

File: Tests/StytchCoreTests/NetworkingClientInterceptor.swift [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                }
+            }
+        }
+    }
+}
+
+private extension Data {
+    func surroundInDataJSONContainer() -> Data {
+        // Surround in a data json container {"data":<existing contents>}
+        var result: [UInt8] = [123, 34, 100, 97, 116, 97, 34, 58, 125]
```
Comment (jhaven-stytch) [inline Tests/StytchCoreTests/NetworkingClientInterceptor.swift:45]: I don't understand whats happening here, help me out?

File: Sources/StytchCore/StytchClient/TOTP/StytchClient+TOTP.swift [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        /// Wraps Stytch's [authenticate](https://stytch.com/docs/api/totp-authenticate) endpoint. Call this method to authenticate a TOTP code entered by a user.
+        public func authenticate(parameters: AuthenticateParameters) async throws -> AuthenticateResponseType {
+            try await router.post(to: .authenticate, parameters: parameters) as AuthenticateResponse
+        }
+
+        /// Wraps Stytch's [recovery_codes](https://stytch.com/docs/api/totp-get-recovery-codes) endpoint. Call this method to retrieve the recovery codes for a TOTP instance tied to a user. Note: If a user has enrolled another MFA method, this method will require MFA. See the [Multi-factor authentication](https://stytch.com/docs/sdks/javascript-sdk#resources_multi-factor-authentication) section for more details.
+        public func recoveryCodes() async throws -> RecoveryCodesResponse {
+            try await router.post(to: .recoveryCodes, parameters: JSON())
```
Comment (jhaven-stytch) [inline Sources/StytchCore/StytchClient/TOTP/StytchClient+TOTP.swift:18]: The API docs say we pass a userId, but this is an empty JSON object, I'm assuming there's some magic somewhere that adds that?

File: Sources/StytchCore/StytchClient/TOTP/StytchClient+TOTP.swift [hunk 3]
```diff
@@ -0,0 +1,125 @@
+public extension StytchClient {
+    /// Time-based One-time Passcodes (TOTPs) are one-time passcodes that are generated based on a shared secret and the current time. TOTPs are also often referred to as Authenticator Apps and are a common form of secondary authentication. Creating a Stytch instance of a TOTP for a User creates a shared secret. This secret is shared by Stytch with the end user's authenticator application of choice (e.g. Google Authenticator). The authenticator app can then generate TOTPs that are valid for a specific amount of time (generally 30 seconds). The end user inputs the TOTP and the developer can use the authenticate method to verify that the TOTP is valid. To call these methods, TOTPs must be enabled in the [SDK Configuration page](https://stytch.com/dashboard/sdk-configuration) of the Stytch dashboard.
+    struct TOTP {
+        let router: NetworkingRouter<TOTPRoute>
+
+        /// Wraps Stytch's [create](https://stytch.com/docs/api/totp-create) endpoint. Call this method to create a new TOTP instance for a user. The user can use the authenticator application of their choice to scan the QR code or enter the secret.
```
Comment (jhaven-stytch) [inline Sources/StytchCore/StytchClient/TOTP/StytchClient+TOTP.swift:6]: Missing the codegen annotations?

Comment (dan-stytch) [reply Sources/StytchCore/StytchClient/TOTP/StytchClient+TOTP.swift:7]: Good catch, thanks!!

File: Sources/StytchCore/StytchClient/TOTP/StytchClient+TOTP.swift (hunk 2 above)
Comment (dan-stytch) [reply Sources/StytchCore/StytchClient/TOTP/StytchClient+TOTP.swift:21]: Web-backend uses the session auth header to get the user id and pass that to the api ([here](https://github.com/stytchauth/web-backend/blob/main/app/sdk/totps.ts#L88)). The expected parameters are [here](https://github.com/stytchauth/web-backend/blob/main/app/sdk/totps.ts#L72)

File: Tests/StytchCoreTests/NetworkingClientInterceptor.swift (hunk 1 above)
Comment (dan-stytch) [reply Tests/StytchCoreTests/NetworkingClientInterceptor.swift:45]: The `Data` type is simply a wrapper around a byte buffer ([docs here](https://developer.apple.com/documentation/foundation/data)) and this is just doing a lower level manipulation of that byte buffer (vs deserializing/reserializing), and adding the existing bytes in-between the uint8 values that correspond to `{"data":` and `}`

Comment (jhaven-stytch) [reply Tests/StytchCoreTests/NetworkingClientInterceptor.swift:45]: Gotcha, that's along the lines of what I assumed (I could at least read "data" out of the code points lol)

**Review by jhaven-stytch — APPROVED** [review]

File: Sources/StytchCore/StytchClient/TOTP/StytchClient+TOTP.swift (hunk 2 above)
Comment (dan-stytch) [reply Sources/StytchCore/StytchClient/TOTP/StytchClient+TOTP.swift:21]: BTW, when you need it, I've laid out the little bit of TOTP info [here](https://docs.google.com/document/d/1OcCESxjfre5ztDm7qp57tbAOZyMyVfqbHlrbwbwXJWQ/edit#heading=h.eqdbtak2xz9l)

**Pull request merged**